### ISSUE TITLE
curation supply.wind_turbine_library

### ISSUE THREAD

**Issue by christian-rli** [OpenEnergyPlatform/data-preprocessing#28]
The [wind_turbine_library](https://openenergy-platform.org/dataedit/view/supply/wind_turbine_library) is a collection of wind turbine data, containing among other things power curves. Data review and metadata creation were discussed in #26 . This issue will remain open for curation purposes. Feel free to comment annotations or point out errors. Feedback will be implemented as soon as possible. 

Low threshold contributions can just be a comment in this issue. 

If you'd like to add several turbines, it would help us if you could to upload your changes to the table model_draft.wind_turbine_library_contributions. This way, we can update the wind_turbine_library with much less effort. There will be a python-script to show how this is done. Alternatively you may (for now) upload a .xlsx-file with your changes in the structure of the following  [file](https://github.com/OpenEnergyPlatform/data-preprocessing/files/3497244/wind_turbine_library_structure.xlsx). (Saving a csv-file in libreoffice as .xlsx works fine. GitHub unfortunately doesn't allow csv uploads)

**Comment by euronion** [comment]
Ok, so if I understand the documentation correctly and want to add the NREL 15 MW offshore wind turbine we talked about over at https://github.com/PyPSA/atlite/issues/54 to OEP, I have to post the .xlsx file here, right?

[IEA-15-240-RWT.xlsx](https://github.com/OpenEnergyPlatform/data-preprocessing/files/4275720/IEA-15-240-RWT.xlsx)

What happens next, do I need to do something else?

**Comment by christian-rli** [comment]
> Ok, so if I understand the documentation correctly and want to add the NREL 15 MW offshore wind turbine we talked about over at [PyPSA/atlite#54](https://github.com/PyPSA/atlite/issues/54) to OEP, I have to post the .xlsx file here, right?

You understood correctly :)

> What happens next, do I need to do something else?

You don't have to do anything else. I'll try to get the update done today and will post here when it's online. In any case it won't happen until some time in the afternoon.

**Comment by Ludee** [comment]
Compare and merge turbines from [`atlite` model](https://github.com/PyPSA/atlite/tree/master/atlite/resources/windturbine) @christian-rli 
Also consider this issue: https://github.com/PyPSA/atlite/issues/54

**Comment by christian-rli** [comment]
Your turbine has been added @euronion 
Thank you for letting me know of the other issues @Ludee . I'm going to have a look at them soon.

**Comment by Bachibouzouk** [comment]
I found some turbines which are not in the wind_turbine_library on this website (https://www.wind-turbine-models.com/turbines/882-senvion-3.2m114) however their impressum states that the data is not public domain and one should ask the manufacturer the right to use their data. I was wondering whether asking manufacturer works and if there is a pitch one can use to not scare them that their data will be "stolen" if we publish them with an open license?

**Comment by beritRLI** [comment]
https://www.htw-berlin.de/hochschule/personen/person/?eid=10251
war vorher bei Senvion
der ist extrem nett; Kollege von RLI-Gründungs-Chef Jochen

**Comment by Bachibouzouk** [comment]
danke @beritRLI  :)

**Comment by christian-rli** [comment]
We actually asked a lawyer on this matter. My notes of his reply read as follows:

- "Solange sie die Originaldaten der Hersteller sammeln, habe ich gar keine Bedenken. Der Hersteller hat ja gar nichts gesammelt, sondern nur gemessen."

To sum up a long session with explanations of different rights that interact with each other in hazy ways: It is very questionable, whether the manufacturer can claim a copyright on the technical description of their product. They are published measurements which neither constitutes a meaningful effort in terms of creating a database (necessary for "Datenbankherstellerrecht") nor does it seem to reach a sufficiently high enough level of originality (necessary for "Urheberrecht").

So technically, in my view you should be allowed to just take individual specification sheets and extract information as you please. Only if you should adapt on a collection of sheets that was done to get a larger overview  "Datenbankherstellerrecht" applies. In fact it does apply to the wind_turbine_library.

All that aside, it is always nice to let the manufacturer know, however. As far as I know we don't have a template text to use for this.

**Comment by fneum** [comment]
@christian-rli I was looking for the 15 MW turbine @euronion posted in https://github.com/OpenEnergyPlatform/data-preprocessing/issues/28#issuecomment-593411988 last year for inclusion in https://openenergy-platform.org/dataedit/view/supply/wind_turbine_library but couldn't find it. Can you check whether it got lost somewhere?

**Comment by christian-rli** [comment]
You're right @fneum , the turbine is missing in the uploaded table. I do recall adding it, but it must have been overwritten in a later upload. Looking into an update I ran into bad documentation unfortunately, hence the late response. 

It seems that this (last) update was done manually with a quick sql command. However these are no longer supported. These days only API uploads are allowed. That's a good decision, but in the case of the wind turbine library that leaves us with a bit of a challenge, since the chosen data types weren't supported with the API last time we tried.

@jh-RLI could you please write an upload script for the library based on a csv download of the current table? If you can, preserve column types. I'll give you the documentation we have available internally. Then we can think about a working system for updating and leave that info here.

**Comment by jh-RLI** [comment]
The complete dataset is now uploaded here: https://openenergy-platform.org/dataedit/view/model_draft/wind_turbine_library

@christian-rli please move the table to supply schema
@fneum the data from https://github.com/OpenEnergyPlatform/data-preprocessing/issues/28#issuecomment-593411988 is available at id:"140" in the table. Note that the table is currently in model_draft schema and will be moved to supply schema soon.

I created a jupyter notebook to upload the data: 

[wind_turbine_lib_upload.zip](https://github.com/OpenEnergyPlatform/data-preprocessing/files/6496143/wind_turbine_lib_upload.zip)

**Comment by fneum** [comment]
@christian-rli did you have a chance to move this to the supply schema yet?

**Comment by christian-rli** [comment]
Just done. Sorry for the wait.

I tried moving the table from `model_draft` to `supply` directly, which failed, because the old table was still there. Then before moving that one to `model_draft` had to delete the upload by @jh-RLI . Then I had to delete it there, then upload the data again (Thank you for providing the upload script!) to finally move it to the supply schema. BTW I added a turbine_id (152) while I was at it

All of this could have been much easier, had there been something like a renaming function. I couldn't find a documentation on anything like that, so I believe there is none. Or am I missing something @jh-RLI ? If no, I'd create an issue for that :)

**Comment by fneum** [comment]
Thanks!

**Comment by jh-RLI** [comment]
@christian-rli by rename function you mean something like "rename table"? At least I'm not aware of one. Would have used the backend DB access or written a script (which probably wouldn't work because the table was already in the restricted schema)

**Comment by birgits** [comment]
@christian-rli there is a turbine with false curve data in the turbine library. The same wind speed appears more than once in the curve data which leads to errors in the windpowerlib. Could you have a look at it? It's the one with turbine_type = IEA-15-240-RWT.

**Comment by christian-rli** [comment]
Thanks for letting me know @birgits . I'll look into it.

**Comment by christian-rli** [comment]
There are indeed quite a few duplicate wind speeds in the mentioned turbine type. Interestingly there are matching power curve values, so maybe the decimal precision of the wind speeds just wasn't enough to account for this. anyway, I removed all duplicate wind speeds, choosing the middle value when there were more than two. Many odd numbered wind speeds remain, but no duplicated:
[supply__wind_turbine_library.csv](https://github.com/OpenEnergyPlatform/data-preprocessing/files/13923922/supply__wind_turbine_library.csv)

**Comment by jh-RLI** [comment]
I can upload it next week.

**Comment by jh-RLI** [comment]
I finally made some time for it tomorrow :) Sorry was quite busy last week.

**Comment by jh-RLI** [comment]
I have updated the data. Please check as I don't know what exact values should have changed.

**Comment by birgits** [comment]
Thank you @jh-RLI! The buggy wind turbine is now correct, but something else must have gone wrong in the course of the upload, because for some wind turbines, there is now a bracket [ too much in the list of wind speeds. See e.g. turbine_type = "E-82/2300". These brackets do not occur in the csv file @christian-rli posted above, so I guess the problem comes from the upload?

**Comment by jh-RLI** [comment]
This is wired. I read the csv with pandas, maybe there was a bug in there or I accidentally added the "]". I will try again :)

**Comment by jh-RLI** [comment]
Hm for some reason when writing the data to the table (in these 3 cases where the random [ is included in the wind speeds column) it just starts over with writing the array. I do not have time to investigate this at the moment. Currently I dont think this bug is in the OEP.

**Comment by birgits** [comment]
Thank you! IDs 97, 82 and 54 are also broken. I will also include a check in the windpowerlib so that the download won't fail in case of faulty turbine data.

**Comment by jh-RLI** [comment]
They are also fixed now.

**Comment by birgits** [comment]
@jh-RLI now it seems like the turbine library is missing completely.. Could you please check?

**Comment by jh-RLI** [comment]
Oops, I forgot to move it back to the supply topic. Now it's back again.

**Comment by birgits** [comment]
Perfect, thank you! All turbine data looks good now :)

**Comment by jh-RLI** [comment]
I have received an email pointing out another error in the library:

> ENO126/4800 turbine has a small bug where several entries are in one column and others are empty.

I will fix this as soon as I have more time :)
